util/noeintr_close_shared.c

```diff
@@ -0,0 +1,212 @@
+
+#include "entropy.h"
+#include "warnp.h"
+
+#include "noeintr.h"
+
+/* Sanity check compiling. */
+#ifndef NOEINTR_CLOSE_PTHREAD
+#error "Do not use this file directly; compile _normal.c or _pthread.c instead"
+#endif
+
+/* Handle multithreading (if applicable). */
+#if NOEINTR_CLOSE_PTHREAD
+#include <pthread.h>
+
+static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
+#endif
+
+#define MAGIC_COOKIE_LEN 32
+
+static int initialized = 0;
+static int s[2] = {-1, -1};
+static uint8_t cookie_write[MAGIC_COOKIE_LEN];
+
+#ifdef NOEINTR_CLOSE_TESTING
+/*
+ * POSIX is silent about the state of fd if close(2) gets EINTR.  For internal
+ * testing, this function randomly selects between the return values which are
+ * permitted by the standard.  For more info, see:
+ * https://www.daemonology.net/blog/2011-12-17-POSIX-close-is-broken.html
+ * https://www.austingroupbugs.net/view.php?id=529
+ */
+static int
+evil_close(int fd)
+{
+
+	switch (random() % 3) {
+	case 0:
+		/* Normal close(). */
+		return (close(fd));
+	case 1:
+		/* No close(), then return EINTR. */
+		errno = EINTR;
+		return (-1);
+	case 2:
+		/* Normal close(), then return EINTR. */
+		close(fd);
+		errno = EINTR;
+		return (-1);
+	}
+
+	/* Should never arrive here. */
+	warn0("evil_close() should not be here");
+	exit(1);
+}
+#endif
+
+static void
+noeintr_close_atexit(void)
+{
+
+	/* Attempt to close s[0] and s[1] using the normal close(). */
+	if (close(s[0]))
+		warnp("close");
+	if (close(s[1]))
+		warnp("close");
+}
+
+static int
+noeintr_close_init(void)
+{
+
+	/* Create a socket pair. */
+	if (socketpair(AF_UNIX, SOCK_STREAM, 0, s)) {
+		warnp("socketpair");
+		goto err0;
+	}
+
+	/* Set up cleanup function. */
+	if (atexit(noeintr_close_atexit)) {
+		warnp("atexit");
+		goto err1;
+	}
+
+	/* Generate a random cookie. */
+	if (entropy_read(cookie_write, MAGIC_COOKIE_LEN)) {
+		warn0("entropy_read");
+		goto err1;
+	}
+
+	/* Write the cookie to s[0]. */
+	if (noeintr_write(s[0], cookie_write, MAGIC_COOKIE_LEN)
+	    != MAGIC_COOKIE_LEN) {
+		warnp("noeintr_write");
+		goto err1;
+	}
+
+	/* We've finished initializing. */
+	initialized = 1;
+
+	/* Success! */
+	return (0);
+
+err1:
+	/* Attempt to close s[0] and s[1] using the normal close(). */
+	if (close(s[0]))
+		warnp("close");
+	if (close(s[1]))
+		warnp("close");
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/**
+ * noeintr_close(fd):
+ * Close the file descriptor ${fd} per the close(2) system call, but handle
+ * EINTR appropriately.
+ */
+int
+noeintr_close(int fd)
+{
+	uint8_t cookie_recv[MAGIC_COOKIE_LEN];
+	ssize_t r;
+#if NOEINTR_CLOSE_PTHREAD
+	int rc;
+#endif
+
+	/* In a multithreaded environment, lock the mutex here. */
+#if NOEINTR_CLOSE_PTHREAD
+	if ((rc = pthread_mutex_lock(&mutex)) != 0) {
+		warn0("pthread_mutex_lock: %s", strerror(rc));
+		goto err0;
+	}
+#endif
+
+	/* Initialize (if needed). */
+	if ((!initialized) && noeintr_close_init()) {
+		warn0("noeintr_close_init");
+		goto err1;
```

Comment:
Clang `-Wthread-safety-analysis` claims that
```
../../../util/noeintr_close_shared.c:211:11: error: mutex 'mutex' is not held on every path through here [-Werror,-Wthread-safety-analysis]
        return (-1);
                 ^
../../../util/noeintr_close_shared.c:139:12: note: mutex acquired here
        if ((rc = pthread_mutex_lock(&mutex)) != 0) {
                  ^
../../../util/noeintr_close_shared.c:130:1: note: Thread warning in function 'noeintr_close'
{
```

I think this is a bogus error.  It goes away if we jump to `err1` (which would unlock the mutex), so perhaps it doesn't realize that `pthread_mutex_lock()` can fail?  I wondered if we should try unlocking the mutex in all error paths, but POSIX says it's undefined behaviour if you unlock and already-unlocked mutex.

(I've seen this error other places that we have mutexes, such as `spiped/lib/dnsthread.c`, but I didn't investigate those in detail.)